code/nebula/neb.cpp

```diff
@@ -689,253 +653,100 @@ float neb2_get_alpha_2shell(float inner_radius, float outer_radius, float magic_
-
-					// set the rotation speed
-					Neb2_cubes[idx1][src][idx2].rot = 0.0f;
-					Neb2_cubes[idx1][src][idx2].rot_speed = frand_range(-max_rotation, max_rotation);
-					Neb2_cubes[src][idx1][idx2].flash = 0.0f;
-				}
-			}
-			break;
-		case 2:
-			for (idx1=0; idx1<Neb2_slices; idx1++) {
-				for (idx2=0; idx2<Neb2_slices; idx2++) {
-					v = &Neb2_cubes[idx1][idx2][src].pt;
-
-					v->xyz.x = h_incw + (ws * (float)idx1) + frand_range(-Nd->wj, Nd->wj);
-					v->xyz.y = h_inch + (hs * (float)idx2) + frand_range(-Nd->hj, Nd->hj);
-					v->xyz.z = h_incd + (ds * (float)src) + frand_range(-Nd->dj, Nd->dj);
-					vm_vec_add2(v, &cube_corner);
-
-					// set the bitmap
-					Neb2_cubes[idx1][idx2][src].bmap = neb2_get_bitmap();
-
-					// set the rotation speed
-					Neb2_cubes[idx1][idx2][src].rot = 0.0f;
-					Neb2_cubes[idx1][idx2][src].rot_speed = frand_range(-max_rotation, max_rotation);
-					Neb2_cubes[src][idx1][idx2].flash = 0.0f;
-				}
-			}
-			break;
-		default:
-			Int3();
-		break;
-	}
+	Neb2_poofs.push_back(new_poof);
 }
 
-// regenerate the player nebula
-void neb2_regen()
+static uint neb_rand_seed = 0;
+
+void upkeep_poofs()
 {
-	int idx;
 	vec3d eye_pos;
-	matrix eye_orient;
-
-	mprintf(("Regenerating local nebula!\n"));
-
-	// get eye position and orientation
 	neb2_get_eye_pos(&eye_pos);
-	neb2_get_eye_orient(&eye_orient);
 
-	// determine the corner of the cube
-	cube_cen = eye_pos;
-
-	// generate slices of the cube
-	for (idx=0; idx<Neb2_slices; idx++) {
-		neb2_gen_slice(0, idx, &cube_cen);
+	neb_rand_seed = 0;
+
+	// make new poofs
+	for (size_t i = 0; i < Poof_info.size(); i++) {
+		if (!poof_is_used(i))
+			continue;
+		poof_info* pinfo = &Poof_info[i];
+		
+		float gen_side_length = (pinfo->view_dist * UPKEEP_DIST_MULT) * 2;
+		float gen_density = pinfo->density * Poof_density_multiplier;
+
+		float poofs_to_gen = gen_side_length * gen_side_length * gen_side_length * gen_density;
+
+		// store the fractional part, take the integer part
+		Poof_accum[i] = modff(Poof_accum[i] + (poofs_to_gen), &poofs_to_gen);
+		for (int j = 0; j < poofs_to_gen; j++) {
+			vec3d pos = eye_pos;
+			vec3d offset = eye_pos / gen_side_length;
+			vec3d rand_pos = vm_well_distributed_rand_vec(neb_rand_seed, &offset);
+			neb_rand_seed++;
+
+			rand_pos.xyz.x *= gen_side_length / 2;
+			rand_pos.xyz.y *= gen_side_length / 2;
+			rand_pos.xyz.z *= gen_side_length / 2;
+
+			pos += rand_pos;
+
+			// we generated poofs in a cube, now keep only those that are within the view sphere, and weren't within the last view sphere
+			// not terribly efficient but very simple
+			if (vm_vec_dist(&eye_pos, &pos) <= (gen_side_length / 2) &&
+				vm_vec_dist(&Poof_last_gen_pos, &pos) > (gen_side_length / 2))
+				new_poof(i, &pos);
+		}
 	}
-}
 
-/*
- * TODO: remove this
-float max_area = 100000000.0f;
-DCF(max_area, "")
-{
-	dc_stuff_float(&max_area);
+	// cull distant poofs
+	if (!Neb2_poofs.empty()) {
```

Comment:
This should really happen before the new poofs are generated. Otherwise, this'll just needlessly iterate over more poofs (namely all that were just generated) which are guaranteed to never need to be culled, as their generating condition is the same for them to not be culled here.

Reply:
Ah, true. (the latter used to not be the case in previous versions)